File: examples/server/00_setup/02_existing_account/go/main.go
// Package main demonstrates loading an existing Self account from storage.
// See README.md for detailed explanations and concepts.
package main

import (
	"fmt"
	"log"
	"os"

	"github.com/joinself/academy/examples/go/common"
	"github.com/joinself/self-go-sdk/account"
	"github.com/joinself/self-go-sdk/event"
)

func main() {
	fmt.Println("📂 Existing Account Loading Example")
	fmt.Println("===================================")

	// Step 1: Create demo account
	fmt.Println("🔧 Creating demo account...")
	originalAccount := createDemoAccount()
	originalAccount.Close()
	fmt.Println("✅ Demo account created and closed")

	// Step 2: Load existing account
	fmt.Println("\n📂 Loading existing account...")
	if !accountExists() {
		fmt.Println("❌ No account storage found")
		return
	}

	selfAccount := loadAccount()
	defer selfAccount.Close()

	// Step 3: Verify and display
	fmt.Println("\n🔍 Verifying account state...")
	verifyAccount(selfAccount)
	displayAccountInfo(selfAccount)

	fmt.Println("\n✅ Account loading demonstration complete!")
}

func createDemoAccount() *account.Account {
	// Clean existing storage
	if accountExists() {
		os.RemoveAll("./storage")
	}

	return common.SetupAccount(common.AccountConfig{
		Callbacks: account.Callbacks{
			OnConnect: func(acc *account.Account) {
				fmt.Println("✅ Demo account connected")
			},
		},
	})
}

func accountExists() bool {
	_, err := os.Stat("./storage")
	return !os.IsNotExist(err)
}

func loadAccount() *account.Account {
	return common.SetupAccount(common.AccountConfig{
		Callbacks: account.Callbacks{
			OnConnect: func(acc *account.Account) {
				fmt.Println("✅ Account reconnected")
			},
			OnMessage: func(selfAccount *account.Account, msg *event.Message) {
				fmt.Printf("📨 Message received\n")
			},
		},
	})
}

func verifyAccount(selfAccount *account.Account) {
	inboxAddress, err := selfAccount.InboxOpen()
	if err != nil {
		log.Printf("⚠️ Inbox access failed: %v", err)
		return
	}

	fmt.Printf("✅ Network: Connected\n")
	fmt.Printf("✅ Identity: %s\n", inboxAddress.String())
	fmt.Printf("✅ Storage: OK\n")
}

func displayAccountInfo(selfAccount *account.Account) {
	inboxAddress, err := selfAccount.InboxOpen()
	if err != nil {
		log.Printf("⚠️ Could not access account info: %v", err)
		return
	}

	fmt.Printf("\n🆔 Account DID: %s\n", inboxAddress.String())
	fmt.Printf("🔐 Status: Ready for connections\n")

	fmt.Println("\n💡 Note: Inbox addresses change between sessions (expected behavior)")
	fmt.Println("   What persists: storage, keys, connections, message history")
}

File: examples/server/00_setup/02_existing_account/java/Main.kt
import com.joinself.selfsdk.kmp.account.Account
import com.joinself.selfsdk.kmp.account.LogLevel
import com.joinself.selfsdk.kmp.account.Target
import java.io.File

fun main() {
    println("📂 Existing Account Loading Example")
    println("===================================")

    // Step 1: Create demo account
    println("🔧 STEP 1: Creating an account for demonstration...")
    val demoAccount = createDemoAccount()
    println("🔄 Closing account to simulate application restart...")
    // Note: Account auto-cleanup in Kotlin/JVM
    demoAccount.destroyAccount()
    println("✅ Account closed (simulating application shutdown)")

    // Step 2: Load existing account
    println("\n📂 STEP 2: Loading existing account from storage...")
    if (!accountExists()) {
        println("❌ No account storage found")
        return
    }

    println("🔍 Looking for existing account storage...")
    println("✅ Found existing account storage directory")
    
    val account = loadExistingAccount()

    // Step 3: Verify and display
    println("\n🔍 STEP 3: Verifying Account State and Persistence")
    println("==================================================")
    verifyAccount(account)

    println("\n📋 STEP 4: Demonstrating Data Persistence")
    println("==========================================")
    displayAccountInfo(account)

    println("\n✅ Account loading demonstration complete!")
}

fun createDemoAccount(): Account {
    // Clean existing storage for demo
    if (accountExists()) {
        File("./storage").deleteRecursively()
    }

    println("🔧 Creating account for loading demonstration...")
    val account = createAccount()
    println("✅ Demo account connected to Self network")
    println("✅ Demo account created successfully!")
    return account
}

fun accountExists(): Boolean {
    return File("./storage").exists()
}

fun loadExistingAccount(): Account {
    val account = createAccount()
    println("✅ Reconnected to Self network")
    println("✅ Account loaded successfully!")
    return account
}

fun createAccount(): Account {
    val storageKey = "276cb6191a345753adb0897c2c0a89370aebf44ef99e612747bee3cd4e757ffa"
        .chunked(2).map { it.toInt(16).toByte() }.toByteArray()

    val account = Account()
    account.configure(
        storagePath = "./storage",
        storageKey = storageKey,
        rpcEndpoint = Target.PRODUCTION_SANDBOX.rpcEndpoint(),
        objectEndpoint = Target.PRODUCTION_SANDBOX.objectEndpoint(),
        messageEndpoint = Target.PRODUCTION_SANDBOX.messageEndpoint(),
        logLevel = LogLevel.WARN,
        onConnect = {
            /* Connection handled in calling functions */
            signal.release()
        },
        onDisconnect = { _ -> },
        onAcknowledgement = { _ -> },
        onError = { _, _ -> },
        onCommit = { _ -> },
        onKeyPackage = { _ -> },
        onWelcome = { _ -> },
        onProposal = { _ -> },
        onMessage = { _ -> },
        onIntegrity = null
    )

    signal.acquire()  // Simple wait for connection
    return account
}

fun verifyAccount(account: Account) {
    val address = Common.getAccountAddress(account)
    
    println("✅ Network connectivity: OK")
    println("✅ Account identity: $address (fresh inbox address)")
    println("✅ Persistence verification successful!")
    println("   • Account storage and configuration preserved")
    println("   • Cryptographic keys and identity maintained")
    println("   • Network connectivity fully restored")
    
    println("\n💡 Important Note About Inbox Addresses:")
    println("   • Inbox addresses are temporary and change between sessions")
    println("   • This is expected Self SDK behavior for security")
}

fun displayAccountInfo(account: Account) {
    val address = Common.getAccountAddress(account)
    
    println("🆔 Account DID: $address")
    println("🔐 Status: Loaded and ready for secure communication")
    
    println("\n💡 Note: Inbox addresses change between sessions (expected behavior)")
    println("   What persists: storage, keys, connections, message history")
}
